### PR TITLE
API V3 Subproject Creation Bug fix

### PR DESCRIPTION
**Pull request by saadmk11**
fixes: https://github.com/readthedocs/readthedocs.org/issues/6271

### CODE REVIEW

File: readthedocs/api/v3/tests/test_subprojects.py [hunk 1]
```diff
@@ -109,7 +110,55 @@ def test_projects_subprojects_list_post_with_subproject_of_itself(self):
             'Project can not be subproject of itself',
             response.json()['non_field_errors'],
         )
-        self.assertEqual(self.project.subprojects.count(), 1)
+        self.assertEqual(newproject.subprojects.count(), 0)
+
+    def test_projects_subprojects_list_post_with_child_already_superproject(self):
+        newproject = self._create_new_project()
+        self.assertEqual(newproject.subprojects.count(), 0)
```
Comment (stsewd) [inline readthedocs/api/v3/tests/test_subprojects.py]:
```suggestion
        self.assertEqual(newproject.subprojects.count(), 0)
        self.assertTrue(self.project.subprojects.exists())
```

File: readthedocs/api/v3/tests/test_subprojects.py [hunk 2]
```diff
@@ -109,7 +110,55 @@ def test_projects_subprojects_list_post_with_subproject_of_itself(self):
...
+        self.assertEqual(response.status_code, 400)
+        self.assertIn(
+            'Child is already a superproject',
+            response.json()['child'],
+        )
+        self.assertEqual(newproject.subprojects.count(), 0)
+
+    def test_projects_subprojects_list_post_with_child_already_subproject(self):
+        newproject = self._create_new_project()
+        self.assertEqual(newproject.subprojects.count(), 0)
```
Comment (stsewd) [inline readthedocs/api/v3/tests/test_subprojects.py]:
```suggestion
        self.assertEqual(newproject.subprojects.count(), 0)
        self.assertTrue(self.project.superprojects.exists())
```

**Comment by saadmk11** [comment]
> Changes look good. I'm leaving some suggestions to have more explicit tests. Also, I'm pushing some tests for the UI form, I was in the process of fixing this when I saw this PR p:
Sure thanks :)

> Oh, and looks like all validation errors from the subproject serializer are missing i18n, can you add it?

Okay :)

File: readthedocs/api/v3/tests/test_subprojects.py [hunk 3]
```diff
@@ -89,17 +89,18 @@ def test_projects_subprojects_list_post_with_others_as_child(self):
         self.assertEqual(self.project.subprojects.count(), 1)
 
     def test_projects_subprojects_list_post_with_subproject_of_itself(self):
-        self.assertEqual(self.project.subprojects.count(), 1)
+        newproject = self._create_new_project()
+        self.assertEqual(newproject.subprojects.count(), 0)
         self.client.credentials(HTTP_AUTHORIZATION=f'Token {self.token.key}')
         data = {
-            'child': self.project.slug,
+            'child': newproject.slug,
```
[the remark this answers is not in the document]
Comment (saadmk11) [reply readthedocs/api/v3/tests/test_subprojects.py]: I'm doing the same thing with a new project and posting `'child': newproject.slug,` to the `newproject` URL. The reason for doing this is `self.project` is already a super project of another project and there is a validation for that in `validate_child()` which gets checked before `validate()` and returns that error. ie: `'Child is already a superproject'`

**Comment by saadmk11** [comment]
> I was almost sure that I was going to make a mistake with this logic. I'm happy that you solved the issue. Thanks!
> 
> When working on this, I asked myself if it wouldn't be better to move the logic outside the Form and share it with the Serializer. After the bug that I introduced, I'd say it's clear that we have to do that instead of replicating the same logic. I want a function `is_valid_as_subproject(project, subproject)` raising different customized exceptions that we can handle properly on the Form and Serializer.

I think we can have 2 methods that will check `is_valid_as_subproject` and `is_valid_as_superproject` or we can have 1 method that checks both like `is_valid_project_relation(project, subproject)`. What do you suggest? :)

Should I update this this PR with that or Its alright for now?

**Comment by humitos** [comment]
> I think we can have 2 methods that will check `is_valid_as_subproject` and `is_valid_as_superproject` or we can have 1 method that checks both like `is_valid_project_relation(project, subproject)`. What do you suggest? :)

Two methods are more explicit to me.

> Should I update this PR with that or Its alright for now?

I think this refactor can be done in another PR. Do you want to work on that as well?

**Comment by saadmk11** [comment]
> I think this refactor can be done in another PR. Do you want to work on that as well?
Sure! :+1: I'll send a PR Soon

**Pull request merged**